### ISSUE TITLE
Proposal: Returning nearest points in world coordinates rather than local cooridnates

### ISSUE THREAD

**Issue by jslee02** [flexible-collision-library/fcl#171]
The reference coordinates of nearest points of distance result are the local coordinates of each corresponding object.

It would make sense to have the same reference coordinate with the contact point, which is the world coordinate, for consistency unless we have a good reason not to do. Either way, I believe we should add comment says what the reference coordinates are.

**Comment by jslee02** [comment]
After further testing, I realized that FCL uses world coordinates for meshes while local coordinates are used for primitive shapes. The existing tests couldn't catch this because they only check distance but not the nearest points properly.

I'll create a pull request for making FCL consistently returns nearest points in world coordinates.

**Comment by wxmerkt** [comment]
@jslee02 Do you have an update on this? (I may have missed the PR). Running off the current master, we noticed that this behaviour still exists.

**Comment by wxmerkt** [comment]
I believe this has been resolved by https://github.com/flexible-collision-library/fcl/pull/288 - can this be closed @SeanCurtis-TRI?

**Comment by SeanCurtis-TRI** [comment]
Yep. I believe we can close this. FTR I want to give @Levi-Armstrong the credit for the work. I just polished and helped resolve CI issues.

**Comment by wxmerkt** [comment]
From a quick test, I am unclear whether these fixes also include primitive-vs-mesh distance queries - which for me report some odd values still (I have a primitive sphere with radius one centred at -1.5 and a BVH sphere with radius one centred at 1.5 - i.e., the closest points should be -0.5 and 0.5 respectively, however, it reports -1 and 1 -- the 1m distance is _correct_ though). I saw the in-code comments that the points for meshes are not guaranteed to be on the surface of the objects - does this apply to be in distance?

**Comment by wxmerkt** [comment]
Actually, I found two very clear (and odd) case of where this is not fixed (with libccd):

1. If ``o1`` is an ``OT_GEOM`` and ``o2`` is an ``OT_BVH``, the ``nearest_points`` are returned in the local frame and in the wrong order (i.e. swapped).
2. If ``o1`` is an ``OT_BVH`` and ``o2`` is an ``OT_GEOM``, the ``nearest_points`` are returned in the local frame (but are in the correct order).

cc: @SeanCurtis-TRI et al.

**Comment by SeanCurtis-TRI** [comment]
@wxmerkt Good catch. I hadn't done the full search for all permutations. And would this relate to your observed issue? Were you spheres tesselated meshes (or one of them)? Certainly, if they were both primitives, your result would *really* surprise me in light of the [tests in this file](https://github.com/flexible-collision-library/fcl/blob/master/test/test_fcl_sphere_sphere.cpp).

**Comment by SeanCurtis-TRI** [comment]
Oops...I just read your note and you explicitly called out primitive-mesh queries.  :)  That's what I get for skimming words and looking for numbers.

**Comment by SeanCurtis-TRI** [comment]
So, commit #288 was correct as far as it went, but didn't go far enough. That's the issue.
